### ISSUE TITLE
Djangos "process_exception()" does not trigger in Middleware from Django Ninja API calls

### ISSUE THREAD

**Issue by MattTPin** [vitalik/django-ninja#698]
Hello,

I'm building a Django project and I'm trying to setup a blanket logging system that will keep a record of API calls performed by my backend.

Looking online I've seen that a way to go about this seems to be to use a combination of middleware and Djangos built into logger. Threads like these [1](https://zeroes.dev/p/django-middleware-to-log-requests/), [2](https://stackoverflow.com/questions/13284890/python-django-automatically-log-when-exceptions-occur-including-request-info), [3](https://gist.github.com/ne0zen/4588a0ceb51f534e2fdb60f513a4fa86), [4](https://stackoverflow.com/questions/67908161/intercepting-a-django-500-error-for-logging-without-creating-serving-a-custom-5), [5](https://stackoverflow.com/questions/6234405/logging-uncaught-exceptions-in-python)

For example (modified from [1](https://zeroes.dev/p/django-middleware-to-log-requests/)) something along these lines...

```
"""
Middleware to log `*/api/*` requests and responses.
"""
import socket
import time
import json
import logging

request_logger = logging.getLogger(__name__)

class RequestLogMiddleware:
    """Request Logging Middleware."""

    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        log_data = {
            "remote_address": request.META["REMOTE_ADDR"],
            "server_hostname": socket.gethostname(),
            "request_method": request.method,
            "status_code": response.status_code,
            "request_path": request.get_full_path(),
        }

        # Only logging "*/api/*" patterns
        if "/api/" in str(request.get_full_path()):
            req_body = json.loads(request.body.decode("utf-8")) if request.body else {}
            log_data["request_body"] = req_body

        # request passes on to controller
        response = self.get_response(request)

        request_logger.info(msg=log_data)

        return response

    # Log unhandled exceptions as well
    def process_exception(self, request, exception): # IS NEVER TRIGGERED WITH DJANGO NINA API CALLS
        try:
            raise exception
        except Exception as e:
            request_logger.exception("Unhandled Exception: " + str(e))
        return exception
```

In settings...
```
MIDDLEWARE = [
    ...
    "django.contrib.messages.middleware.MessageMiddleware",
    # Request Logger
    "<application>.middleware.request_log.RequestLogMiddleware",
]
```

I've set this up for my project and it's working great for API calls that run without any exceptions, however, catching of unhandled exceptions by [_process_exception()_](https://docs.djangoproject.com/en/4.1/topics/http/middleware/#process-exception) in _RequestLogMiddleware()_ does not seem to be working as expected. This is because it does not seem to be invoked while making API calls using Django Ninja that I intentionally make throw unhandled exceptions.

I'm wondering if perhaps Django Ninja handles its exceptions in a way that prevents _process_exception()_ from being invoked. Is there a setting I can adjust that will allow these exceptions in Django Ninja API calls to trigger _process_exception()_?

I am able to check the status code of each call which has allowed me to detect when errors occur and log them from within the __call__ portion of the middleware, however, I am not able to find a way to get access to the traceback information (which is what I am primarily interested in logging). 

For greater context here is how other parts of my project are configured...

urls.py
```
urlpatterns = [
    path("api/admin/", admin.site.urls),
    path("api/", api.urls),
]

if settings.DEBUG:
    urlpatterns += static(settings.STATIC_URL, document_root=settings.STATIC_ROOT)
    urlpatterns += static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)
```

secondary_api_file.py
```
router = Router()

@router.post("/accept_article_single")
def accept_article_single(request,) -> list:
    API Call functionality...
```

api.py
```
from .secondary_api_file import (router as secondary_api_file_router,)

api.add_router("", secondary_api_file_router)
```

In summary either of the following would solve my issue...

1. Find a way to ensure _process_exception()_ is invoked whenever an API call made through Django Ninja throws an unhandled exception.
2. Find a way to include the unhandled exception stack traceback in my log_data whenever an erroneous status code is identified in the __call__ portion of my middleware.

Perhaps the entire methodology behind my approach is flawed, in which case, any suggestions would be greatly appreciated.

Thank you!

**Comment by vitalik** [comment]
Hi @MattTPin 

by default if settings.DEBUG is True django ninja just outputs traceback to http response (so it does not go to process_exception handler in that case) but if DEBUG is False - then you will get process_exception 

This is done to easier debugging with swagger UI (as by default it dumps html)

see the code that handles exception here - https://github.com/vitalik/django-ninja/blob/master/ninja/errors.py#L100-L108

docs - https://django-ninja.rest-framework.com/guides/errors/#exception

if you want to override this behaviour you can change default exception handler to always raise error:

```Python

@api.exception_handler(Exception)
def global_error(request, exc):
    raise exc

```
